File: frontend/types/coscientist.ts
/**
 * Co-Scientist 类型定义 — 与后端 app/schemas/coscientist.py 对齐
 *
 * 多智能体科学推理引擎的类型契约：
 * - 运行管理（RunCreate/RunResponse）
 * - 假设排名（RankedHypothesis）
 * - 辩论日志（DebateLog）
 * - 进化树（EvolutionNode/Edge）
 * - 专家反馈（FeedbackPayload）
 * - WebSocket 事件（CoScientistWSEvent）
 */

// ========== 运行状态枚举 ==========

export type RunStatus =
  | 'pending'
  | 'running'
  | 'awaiting_feedback'
  | 'completed'
  | 'failed'
  | 'cancelled';

export type RunPhase =
  | 'generation'
  | 'proximity'
  | 'reflection'
  | 'debate'
  | 'ranking'
  | 'evolution'
  | 'meta_review';

// 验证案例（aml/liver_fibrosis/amr）已删除，仅保留 custom
export type CaseType = 'custom';

export type FeedbackType =
  | 'constraint'
  | 'approval'
  | 'rejection'
  | 'new_evidence'
  | 'directional'
  | 'veto'
  | 'elo_adjustment'
  | 'refinement';

// ========== 运行 ==========

export interface RunCreate {
  research_goal: string;
  project_id?: string;
  case_type?: CaseType;
  max_rounds?: number;
  initial_hypothesis_count?: number;
  config?: Record<string, unknown>;
}

export interface RunResponse {
  id: string;
  user_id: string;
  project_id?: string | null;
  session_id?: string | null;
  research_goal: string;
  case_type?: string | null; // string 而非 CaseType，兼容历史记录中的旧 case_type 值
  status: RunStatus;
  current_round: number;
  max_rounds: number;
  current_phase?: RunPhase | null;
  config?: Record<string, unknown> | null;
  final_rankings?: Record<string, unknown> | null;
  meta_review?: string | null;
  expert_feedback?: Array<Record<string, unknown>> | null;
  total_cost_usd?: number | null;
  duration_sec?: number | null;
  started_at?: string | null;
  completed_at?: string | null;
  error_message?: string | null;
  created_at: string;
  updated_at: string;
}

export interface RunListResponse {
  items: RunResponse[];
  total: number;
  page: number;
  page_size: number;
}

// ========== 假设排名 ==========

export interface RankedHypothesis {
  id: string;
  name: string;
  description?: string | null;
  mechanism?: string | null;
  elo_score: number;
  novelty_score?: number | null;
  plausibility_score?: number | null;
  testability_score?: number | null;
  safety_score?: number | null;
  rank?: number | null;
  evolution_strategy?: string | null;
  parent_ids?: string[] | null;
  critique_summary?: string | null;
  status: string;
  experimental_elo_adjustment?: number | null;
  experimental_validation_count?: number | null;
}

export interface RankingsResponse {
  run_id: string;
  round_num: number;
  rankings: RankedHypothesis[];
  total_hypotheses: number;
}

// ========== 辩论日志 ==========

export interface DebateLog {
  id: string;
  run_id: string;
  hypothesis_id: string;
  round_num: number;
  proponent_argument: string;
  opponent_argument: string;
  judge_assessment?: string | null;
  consensus_score?: number | null;
  mechanism_agreed?: boolean | null;
  refined_hypothesis?: string | null;
  cost_usd?: number | null;
  created_at: string;
}

export interface DebateListResponse {
  run_id: string;
  debates: DebateLog[];
  total: number;
}

// ========== 进化树 ==========

export interface EvolutionNode {
  hypothesis_id: string;
  name: string;
  evolution_strategy: string;
  parent_ids: string[];
  elo_score: number;
  round_num: number;
  rank?: number | null;
}

export interface EvolutionEdge {
  from_id: string;
  to_id: string;
  strategy: string;
}

export interface EvolutionTreeResponse {
  run_id: string;
  nodes: EvolutionNode[];
  edges: EvolutionEdge[];
  total_rounds: number;
}

// ========== Meta-review ==========

export interface MetaReviewResponse {
  run_id: string;
  meta_review: string;
  final_rankings?: Record<string, unknown> | null;
  total_cost_usd?: number | null;
  duration_sec?: number | null;
  completed_at?: string | null;
}

// ========== 专家反馈 ==========

export interface FeedbackPayload {
  feedback_text: string;
  feedback_type: FeedbackType;
  target_hypothesis_id?: string;
}

export interface FeedbackResponse {
  accepted: boolean;
  message: string;
  applied_round?: number | null;
  parsed_constraints?: Record<string, unknown> | null;
}

// ========== 案例 ==========

export interface CaseInfo {
  case_type: CaseType;
  name: string;
  description: string;
  research_goal_template: string;
  expected_benchmarks: Record<string, unknown>;
}

export interface CaseListResponse {
  cases: CaseInfo[];
}

// ========== AI 智能生成研究目标 ==========

export interface GenerateGoalResult {
  research_goal: string;
  suggested_case_type: string | null;
  suggested_max_rounds: number;
  suggested_initial_count: number;
  framework: string[];
  key_questions: string[];
  content_suggestions: string[];
}

// ========== 综合性研究模板 ==========

export interface ComprehensiveTemplate {
  case_type: string;
  name: string;
  description: string;
  research_goal_template: string;
  expected_benchmarks: Record<string, any>;
  sub_templates: Array<{ case_type: string; name: string; description: string }>;
  framework: string[];
  config_presets: Record<string, { max_rounds: number; initial_count: number; description: string }>;
}


// ========== Agent 活动 ==========

export interface AgentActivity {
  agent_name: string;
  status: 'idle' | 'running' | 'completed' | 'failed';
  current_task?: string | null;
  started_at?: string | null;
  completed_at?: string | null;
  token_usage?: Record<string, number> | null;
  cost_usd?: number | null;
  error?: string | null;
}

export interface AgentActivityFeedResponse {
  run_id: string;
  agents: AgentActivity[];
  current_phase?: RunPhase | null;
  current_round: number;
}

// ========== 进度 ==========

export interface ProgressSnapshot {
  run_id: string;
  status: RunStatus;
  current_round: number;
  max_rounds: number;
  current_phase?: RunPhase | null;
  recent_events: WSEventPayload[];
  total_cost_usd?: number;
  total_tokens?: number;
}

// ========== WebSocket 事件 ==========

export interface WSEventPayload {
  type: string;
  run_id: string;
  timestamp: string;
  payload: Record<string, unknown>;
}

export interface WSClientMessage {
  type: 'subscribe' | 'unsubscribe' | 'feedback' | 'cancel' | 'ping';
  run_id?: string;
  payload?: Record<string, unknown>;
}

File: backend/app/schemas/coscientist.py
"""Co-Scientist Pydantic Schemas - 请求/响应模型

基于 Nature 论文 Co-Scientist 多智能体科学推理引擎的 API 契约。
"""
from datetime import datetime
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.coscientist_run import CaseType, RunPhase, RunStatus


# ========== 运行创建 ==========

class RunCreate(BaseModel):
    """提交研究目标，启动 Co-Scientist 运行"""
    research_goal: str = Field(..., min_length=10, max_length=5000, description="自然语言研究目标")
    project_id: Optional[UUID] = Field(None, description="关联项目 ID")
    case_type: Optional[str] = Field(None, description="案例类型: custom（自定义研究目标）。历史值 aml/liver_fibrosis/amr 已删除，仅作为字符串兼容旧记录")
    max_rounds: int = Field(5, ge=1, le=10, description="最大迭代轮数")
    initial_hypothesis_count: int = Field(5, ge=3, le=10, description="初始假设数量")
    config: Optional[Dict[str, Any]] = Field(None, description="运行配置覆盖")


class RunResponse(BaseModel):
    """运行详情响应"""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: UUID
    project_id: Optional[UUID] = None
    session_id: Optional[UUID] = None
    research_goal: str
    case_type: Optional[str] = None
    status: str = RunStatus.PENDING
    current_round: int = 0
    max_rounds: int = 5
    current_phase: Optional[str] = None
    config: Optional[Dict[str, Any]] = None
    # final_rankings 实际为排名后的假设列表（List），非字典；用 Any 兼容 list/dict/None
    final_rankings: Optional[Any] = None
    meta_review: Optional[str] = None
    expert_feedback: Optional[List[Dict[str, Any]]] = None
    total_cost_usd: Optional[float] = None
    duration_sec: Optional[float] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    error_message: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class RunListResponse(BaseModel):
    """运行列表分页响应"""
    items: List[RunResponse]
    total: int
    page: int = 1
    page_size: int = 20


# ========== 专家反馈 ==========

class FeedbackPayload(BaseModel):
    """专家反馈请求"""
    feedback_text: str = Field(..., min_length=1, max_length=10000, description="反馈文本")
    feedback_type: str = Field(
        ...,
        description="反馈类型: constraint(约束)/approval(认可)/rejection(否决)/new_evidence(新证据)",
    )
    target_hypothesis_id: Optional[UUID] = Field(None, description="可选：针对特定假设")


class FeedbackResponse(BaseModel):
    """反馈提交响应"""
    accepted: bool
    message: str
    applied_round: Optional[int] = None
    parsed_constraints: Optional[Dict[str, Any]] = None


# ========== 假设排名 ==========

class RankedHypothesisView(BaseModel):
    """排名假设视图"""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    description: Optional[str] = None
    mechanism: Optional[str] = None
    elo_score: float = 1000.0
    novelty_score: Optional[float] = None
    plausibility_score: Optional[float] = None
    testability_score: Optional[float] = None
    safety_score: Optional[float] = None
    rank: Optional[int] = None
    evolution_strategy: Optional[str] = None
    parent_ids: Optional[List[str]] = None
    critique_summary: Optional[str] = None
    status: str = "draft"
    experimental_elo_adjustment: Optional[float] = None
    experimental_validation_count: Optional[int] = None


class RankingsResponse(BaseModel):
    """排名响应"""
    run_id: UUID
    round_num: int
    rankings: List[RankedHypothesisView]
    total_hypotheses: int


# ========== 辩论日志 ==========

class DebateLogView(BaseModel):
    """辩论日志视图"""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    run_id: UUID
    hypothesis_id: UUID
    round_num: int
    proponent_argument: str
    opponent_argument: str
    judge_assessment: Optional[str] = None
    consensus_score: Optional[float] = None
    mechanism_agreed: Optional[bool] = None
    refined_hypothesis: Optional[str] = None
    cost_usd: Optional[float] = None
    created_at: datetime


class DebateListResponse(BaseModel):
    """辩论日志列表响应"""
    run_id: UUID
    debates: List[DebateLogView]
    total: int


# ========== 进化树 ==========

class EvolutionNode(BaseModel):
    """进化树节点"""
    hypothesis_id: str
    name: str
    evolution_strategy: str
    parent_ids: List[str] = Field(default_factory=list)
    elo_score: float = 1000.0
    round_num: int = 0
    rank: Optional[int] = None


class EvolutionEdge(BaseModel):
    """进化树边"""
    from_id: str
    to_id: str
    strategy: str


class EvolutionTreeResponse(BaseModel):
    """进化树响应"""
    run_id: UUID
    nodes: List[EvolutionNode]
    edges: List[EvolutionEdge]
    total_rounds: int


# ========== Meta-review ==========

class MetaReviewResponse(BaseModel):
    """Meta-review 报告响应"""
    run_id: UUID
    meta_review: str
    final_rankings: Optional[Any] = None
    total_cost_usd: Optional[float] = None
    duration_sec: Optional[float] = None
    completed_at: Optional[datetime] = None


# ========== 案例 ==========

class CaseInfo(BaseModel):
    """案例信息"""
    case_type: str
    name: str
    description: str
    research_goal_template: str
    expected_benchmarks: Dict[str, Any]


class CaseListResponse(BaseModel):
    """案例列表响应"""
    cases: List[CaseInfo]


class CaseRunRequest(BaseModel):
    """案例运行请求"""
    project_id: Optional[UUID] = None
    max_rounds: int = Field(3, ge=1, le=10)
    custom_goal: Optional[str] = Field(None, max_length=5000, description="自定义研究目标（覆盖案例预设）")


# ========== WebSocket 事件 ==========

class CoScientistWSEvent(BaseModel):
    """Co-Scientist WebSocket 推送事件信封

    type 取值：
    - round_started: 轮次开始 {round_num, phase}
    - generation_done: 假设生成完成 {hypotheses}
    - debate_round: 辩论回合 {hypothesis_id, round, proponent, opponent, consensus}
    - elo_update: Elo 评分更新 {hypothesis_id, old_elo, new_elo, match_result}
    - evolution: 假设进化 {parent_ids, child_id, strategy}
    - ranking_updated: 排名更新 {ranked_list}
    - feedback_required: 需要专家反馈 {reason, current_hypotheses}
    - meta_review_done: Meta-review 完成 {report}
    - run_completed: 运行完成 {final_rankings}
    - error: 错误 {error, code}
    """
    type: str
    run_id: str
    timestamp: str
    payload: Dict[str, Any] = Field(default_factory=dict)


class CoScientistWSClientMessage(BaseModel):
    """Co-Scientist WebSocket 客户端消息

    type 取值：
    - subscribe: 订阅运行进度
    - unsubscribe: 取消订阅
    - feedback: 提交专家反馈
    - cancel: 取消运行
    - ping: 心跳
    """
    type: str
    run_id: Optional[str] = None
    payload: Dict[str, Any] = Field(default_factory=dict)


# ========== Agent 活动状态 ==========

class AgentActivity(BaseModel):
    """单个 Agent 活动状态"""
    agent_name: str
    status: str = "idle"  # idle/running/completed/failed
    current_task: Optional[str] = None
    started_at: Optional[str] = None
    completed_at: Optional[str] = None
    token_usage: Optional[Dict[str, int]] = None
    cost_usd: Optional[float] = None
    error: Optional[str] = None


class AgentActivityFeedResponse(BaseModel):
    """Agent 活动流响应"""
    run_id: UUID
    agents: List[AgentActivity]
    current_phase: Optional[str] = None
    current_round: int = 0


# ========== AI 智能生成研究目标 ==========

class GenerateGoalRequest(BaseModel):
    """AI 生成研究目标请求"""
    topic: str = Field(..., min_length=2, max_length=500, description="研究主题关键词或描述")
    project_id: Optional[UUID] = Field(None, description="关联项目 ID（用于上下文感知）")
    case_type: Optional[str] = Field(None, description="可选：指定案例类型风格")


class GenerateGoalResponse(BaseModel):
    """AI 生成研究目标响应"""
    research_goal: str = Field(..., description="生成的研究目标文本（可直接用于创建运行）")
    suggested_case_type: Optional[str] = Field(None, description="建议的案例类型")
    suggested_max_rounds: int = Field(3, description="建议最大迭代轮数")
    suggested_initial_count: int = Field(5, description="建议初始假设数量")
    framework: List[str] = Field(default_factory=list, description="研究框架要点")
    key_questions: List[str] = Field(default_factory=list, description="关键科学问题")
    content_suggestions: List[str] = Field(default_factory=list, description="内容建议")